### ISSUE TITLE
Kops 1.5.3 to 1.6.2 and cluster upgrade problems - Master node does not become ready

### ISSUE THREAD

**Issue by kaspernissen** [kubernetes/kops#2827]
Sorry, if this is a duplicate. I found similar issues [#2806](https://github.com/kubernetes/kops/issues/2806), but then not completely the same, therefore this new issue. 

I'm running into an issue when trying to upgrade from Kops 1.5.3 to 1.6.2. With corresponding Kubernetes versions 1.5.7 to 1.6.2. 

After running a rolling-update, 1 master node will not join the cluster again. 

*Steps to reproduce*

1. Download kops version 1.5.3 and spin up a cluster with the following configuration
```
$ kops create cluster \
  --name $KOPS_NAME \
  --state $KOPS_STATE_STORE \
  --node-count 3 \
  --zones eu-west-1a,eu-west-1b,eu-west-1c \
  --master-zones eu-west-1a,eu-west-1b,eu-west-1c \
  --dns-zone=${PRIVATE_HOSTED_ZONE_ID} \
  --dns private \
  --node-size t2.medium \
  --master-size t2.small \
  --topology private \
  --networking weave \
  --bastion
```
This will bring up a working kubernetes cluster with 3 master nodes, 1 in each AZ as described in the config. 
2. Download and update local kops version to 1.6.2. 
3. Run `kubectl upgrade cluster` - all nodes except one of the master nodes becomes Ready. 
![cluster_update](https://user-images.githubusercontent.com/4429108/27636374-94fa41d4-5c0b-11e7-94fe-be236d2b7bfc.png)

![kubectl_get_nodes_master_not_ready](https://user-images.githubusercontent.com/4429108/27636206-a2be4b9a-5c0a-11e7-82a7-4d777e4cddbb.png)

I tried looking into the logs and found the following:

*daemon.log*
```
Jun 28 11:23:06 ip-172-20-127-106 docker[1550]: I0628 11:23:06.576291      97 addons.go:112] Skipping version range "<1.6.0" that does not match current version 1.6.2
Jun 28 11:23:06 ip-172-20-127-106 docker[1550]: I0628 11:23:06.576342      97 addons.go:112] Skipping version range "<1.6.0" that does not match current version 1.6.2
Jun 28 11:23:06 ip-172-20-127-106 docker[1550]: I0628 11:23:06.576355      97 addons.go:112] Skipping version range "<1.6.0" that does not match current version 1.6.2
Jun 28 11:23:06 ip-172-20-127-106 docker[1550]: No update required
Jun 28 11:23:07 ip-172-20-127-106 kubelet[1619]: W0628 11:23:07.803004    1619 cni.go:157] Unable to update cni config: No networks found in /etc/cni/net.d/
Jun 28 11:23:07 ip-172-20-127-106 kubelet[1619]: E0628 11:23:07.803603    1619 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 28 11:23:12 ip-172-20-127-106 kubelet[1619]: W0628 11:23:12.808564    1619 cni.go:157] Unable to update cni config: No networks found in /etc/cni/net.d/
Jun 28 11:23:12 ip-172-20-127-106 kubelet[1619]: E0628 11:23:12.809170    1619 kubelet.go:2067] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Jun 28 11:23:13 ip-172-20-127-106 kubelet[1619]: I0628 11:23:13.440673    1619 aws.go:937] Could not determine public DNS from AWS metadata.
```

*etcd.log*
```
2017-06-28 11:08:06.572336 I | rafthttp: the connection with 658e948eb50fb1e9 became active
2017-06-28 11:08:06.572431 I | rafthttp: the connection with 117cb38d1787f2f1 became active
2017-06-28 11:08:06.657337 I | raft: raft.node: ac1a9daa5ac57018 elected leader 117cb38d1787f2f1 at term 32
2017-06-28 11:08:06.698412 I | etcdserver: published {Name:etcd-c ClientURLs:[http://etcd-c.internal.k8s.test.REMOVED-NAME.com:4001]} to cluster a18bb12a913e19df
2017-06-28 11:12:04.322918 I | raft: ac1a9daa5ac57018 is starting a new election at term 32
2017-06-28 11:12:04.322963 I | raft: ac1a9daa5ac57018 became candidate at term 33
2017-06-28 11:12:04.322983 I | raft: ac1a9daa5ac57018 received vote from ac1a9daa5ac57018 at term 33
2017-06-28 11:12:04.322994 I | raft: ac1a9daa5ac57018 [logterm: 32, index: 7318] sent vote request to 117cb38d1787f2f1 at term 33
2017-06-28 11:12:04.323001 I | raft: ac1a9daa5ac57018 [logterm: 32, index: 7318] sent vote request to 658e948eb50fb1e9 at term 33
2017-06-28 11:12:04.323009 I | raft: raft.node: ac1a9daa5ac57018 lost leader 117cb38d1787f2f1 at term 33
2017-06-28 11:12:04.332300 I | raft: ac1a9daa5ac57018 received vote from 658e948eb50fb1e9 at term 33
2017-06-28 11:12:04.332318 I | raft: ac1a9daa5ac57018 [q:2] has received 2 votes and 0 vote rejections
2017-06-28 11:12:04.332335 I | raft: ac1a9daa5ac57018 became leader at term 33
2017-06-28 11:12:04.332346 I | raft: raft.node: ac1a9daa5ac57018 elected leader ac1a9daa5ac57018 at term 33
2017-06-28 11:12:04.501630 I | raft: ac1a9daa5ac57018 [term: 33] ignored a MsgApp message with lower term from 117cb38d1787f2f1 [term: 32]
2017-06-28 11:12:05.033502 I | raft: ac1a9daa5ac57018 [term: 33] ignored a MsgHeartbeat message with lower term from 117cb38d1787f2f1 [term: 32]
2017-06-28 11:12:05.033531 I | raft: ac1a9daa5ac57018 [term: 33] ignored a MsgHeartbeat message with lower term from 117cb38d1787f2f1 [term: 32]
2017-06-28 11:12:05.538178 W | etcdserver: failed to reach the peerURL(http://etcd-a.internal.k8s.test.REMOVED-NAME.com:2380) of member 117cb38d1787f2f1 (Get http://etcd-a.internal.k8s.test.REMOVED-NAME.com:2380/version: dial tcp 172.20.40.60:2380: i/o timeout)
2017-06-28 11:12:05.538203 W | etcdserver: cannot get the version of member 117cb38d1787f2f1 (Get http://etcd-a.internal.k8s.test.REMOVED-NAME.com:2380/version: dial tcp 172.20.40.60:2380: i/o timeout)
```

*controller-manager.log*
![controller_manager](https://user-images.githubusercontent.com/4429108/27636331-57f757f4-5c0b-11e7-9057-347ade8dae05.png)

I found this issues: https://github.com/kubernetes/kubernetes/issues/43815 which is similar, but should only effect 1.6.0.

**Comment by kaspernissen** [comment]
I just tried a clean install using Kops 1.6.2 instead of upgrading with the same configuration as before, and everything works fine. 

I noticed in this newly created cluster all nodes run a weave-net container, whereas in the previous trial, only worker nodes had a running weave-net container. 

I will retry upgrading again tomorrow.

**Comment by kaspernissen** [comment]
This issue [#2377](https://github.com/kubernetes/kops/issues/2377) is also similar. I will give that solution a try tomorrow.

**Comment by mrIncompetent** [comment]
I ran into the same situation. So far i've not been able to fix the issue. Recreating the cluster is not an option.

@kaspernissen any chance to hear from your solution?

**Comment by kaspernissen** [comment]
I'm planning on doing a write up of my experiences, but here comes a quick overview of the steps I did in order to complete the upgrade without issues. 

I did a lot of testing different ways of upgrading. First I followed the describe procedure of doing a rolling-update of the master instances, however, one master never became ready. Then i ran into master nodes not getting the correct Taints on nodes. So I decided to take a more drastic approach of terminating all masters at the same time, after uploading the new configuration, and that actually fixed all my previous issues. However, this seems like a bad tip to promote, but it worked for me. This issue is also describe to some extend here: [2674](urhttps://github.com/kubernetes/kops/issues/2674l). The problem as far as I understand it, is that the elected master is the last 1.5.2 instances back when doing a rolling upgrade. The new master nodes registers with different kubernetes versions, causing som wierd effects. So here is what I did (have tried it at least 3 times with no problems):

**Steps**
First upgrade the cluster:
```
$ kops upgrade cluster
```
Verify that the output is correct. Then run:
```
$ kops upgrade cluster --yes
```

_Applying the configuration to AWS_
Next, we have to update the cluster configuration
```
$ kops update cluster <CLUSTER>
```
Verify and then do:
```
$ kops update cluster<CLUSTER> --yes
```

_Fix the kube-dns configmap_

Does the configmap exist?
```
$ kubectl -n kube-system get configmap kube-dns
```

If not, then create it:
```
$ kubectl create configmap -n kube-system kube-dns
```

_Killing the master nodes_

After a lot of testing, the best solution to avoid the problem of masters in different versions is to kill them using the AWS console at the same time. This may seem a bit risci, but doing rolling-upgrades results in many painpoints and wierd cluster state as mentioned.

```
GO TO AWS CONSOLE AND MARK ALL 3 MASTERS AND THEN TERMINATE
```
Wait for the master to come back up. 

_Rolling-update the worker nodes_
Next, verify that a rolling-update is neeeded:
```
$ kops rolling-update cluster
```
All master nodes should respond with Status: Ready 

Set the flag DrainAndValidateRollingUpdate if you want worker nodes to be drained.

```
$ export KOPS_FEATURE_FLAGS="+DrainAndValidateRollingUpdate"
```

and then execute the rolling-update
```
$ kops rolling-update cluster --yes
```

Then wait until all nodes are updated. 

Hope this helps. This again, is probably not the recommend way of doing it. But I couldn't find a better solution. If you can trigger a master elect of some sort, then this could probably be a more feasable solution.

**Comment by chrislovecnm** [comment]
@kaspernissen this is actually very very good information.  I had not thought of killing all the masters at the same time.  Could be a strategy, for upgrade, if you can have control plane downtime.

Now here is the tricky problem, how do we keep the masters up and guarantee that etcd keeps happy?

Would you agree that
 
```
2017-06-28 11:12:05.538178 W | etcdserver: failed to reach the peerURL(http://etcd-a.internal.k8s.test.REMOVED-NAME.com:2380) of member 117cb38d1787f2f1 (Get http://etcd-a.internal.k8s.test.REMOVED-NAME.com:2380/version: dial tcp 172.20.40.60:2380: i/o timeout)
2017-06-28 11:12:05.538203 W | etcdserver: cannot get the version of member 117cb38d1787f2f1 (Get http://etcd-a.internal.k8s.test.REMOVED-NAME.com:2380/version: dial tcp 172.20.40.60:2380: i/o timeout)
```

Seems to be the crux of our issue?  Was the record not updated?  I am wondering if we are having a problem where the `dns-controller` is not updating DNS properly or is the problem that etcd gets into an odd state. 

Another question, at anypoint in the process did the dns-controller, or any pod in kube-system not get scheduled?  I am wondering if we are hitting capcity issues with t2s.

**Comment by kaspernissen** [comment]
Yes, this will, of course, result in control plane downtime, I forgot to mention that in my previous response. 

@chrislovecnm that is a good question. I was running into a couple of different scenarios when running the rolling-update command. One of the times, the dns-controller was not able to start. It seemed that the dns-controller was not updated and was missing the PodToleration NoSchedule on master nodes. Adding that toleration, made the cluster come up. One time the DNS for the internal api was not updated correctly with the new IPs of the master nodes. These problems are what lead me to try out killing the masters at the same time as a last resort. My thought was, that this would be the same as updating a cluster with one master where the control plane also suffers from downtime. 

I didn't notice any problems with the strategy of killing the masters at the same time. Another small pain point in the rolling-update process with the experimental feature flag is the fact that Kops cordon, drain and kill the node first. Would be great, if possible, to spin up the new node, then cordon, drain, and kill. I know this would probably require a bit more logic, in terms of tweaking the size of the cluster by forcing the auto scaling group to spin up a new instance.

**Comment by chrislovecnm** [comment]
> Would be great, if possible, to spin up the new node, then cordon, drain, and kill. I know this would probably require a bit more logic, in terms of tweaking the size of the cluster by forcing the auto scaling group to spin up a new instance

I have a PR in the works that does this for nodes, as you actually cannot do this with masters ... at least that I know of.

I have seen reports when the dns-controller or say weave does not get updated properly.  Do you have steps to consistently re-create this, or are these intermittent problems?

**Comment by kaspernissen** [comment]
Sounds great, @chrislovecnm.

Unfortunately, I don't have a consistent way to re-create the issue of the dns-controller. I only ran into this once. The original issue with a master not becoming ready was however fairly consistent, however, one time it completed the upgrade without any problems. 

I put some notes about the steps here: [direct upgrade from Kops 1.5.1 to Kops 1.6.2](https://github.com/kaspernissen/kops-upgrade-directly) and [step-by-step](https://github.com/kaspernissen/kops-upgrade-step-for-step)

I actually just noticed that the issue with the dns-controller is present in the step-by-step experiment (look at the bottom). I saved some of the output. 

I can give it a try in the beginning of next week again.
